README.md
service.autosubs
===============

###What is AutoSubs?
Auto run XBMC Subtitles if no subs are found for a Movie/TV Show

###Installation
Exclude media paths that include certain words that could trigger the script in addon settings.

Note:
Some additional configuration settings that can be done in XBMC Subtitles to reduce end-user actions are:
Enable the following in Advanced Options for XBMC Subtitles.
- Search next available service if no results are found
- Auto download first " sync" subtitle

###Thanks
- Icon credits goes to zNET Computer Solutions http://www.znetcs.pl
- ezechiel1917@script.trakt for better way of handling exlusions


changelog.txt
1.1.2 - ghostxwheel
- Fixed: Syntax fix.

1.1.1 - juokelis
- Fixed: Exception settings are ignored

1.1.0 - juokelis
- Ported to python 3 to run on v19 Matrix
- strings.xml > strings.po

1.0.0 - schumi2004,CaTz, Mafarricos
- Several fixes
- Clean up
- Added translations: Dutch, Portuguese
- Exclude setting based on video duration
- Check for specific language

0.0.9 - schumi2004
- added debug function
- added exclude paths

0.0.8 - by sagiben
- popup subtitles window only for Movie files, otherwise it pops also for audio

0.0.7 - by CaTz
- Added feature, popup subtitles window if there is no subtitles for specific language

0.0.6 - by CaTz
- Change the dependency to support Gotham

0.0.5
- Renamed script to service.autosubs

0.0.4
- Added configuration options

0.0.3
- Disabled LiveTV by default

0.0.2
- initial release


resources/lib/autosubs.py
#!/usr/bin/python

import os
import xbmc
import xbmcvfs
import xbmcaddon

__addon__ = xbmcaddon.Addon()
__author__ = __addon__.getAddonInfo('author')
__scriptid__ = __addon__.getAddonInfo('id')
__scriptname__ = __addon__.getAddonInfo('name')
__version__ = __addon__.getAddonInfo('version')
__language__ = __addon__.getLocalizedString
debug = __addon__.getSetting("debug")
__cwd__ = xbmcvfs.translatePath(__addon__.getAddonInfo('path'))
__profile__ = xbmcvfs.translatePath(__addon__.getAddonInfo('profile'))
__resource__ = xbmcvfs.translatePath(os.path.join(__cwd__, 'resources'))

__settings__ = xbmcaddon.Addon("service.autosubs")

ignore_words = (__settings__.getSetting('ignore_words').split(','))
ExcludeTime = int((__settings__.getSetting('ExcludeTime'))) * 60

currentPlaying = ""

monitor = xbmc.Monitor()


def Debug(msg, force=False):
    if debug == "true" or force:
        xbmc.log("#####[AutoSubs]##### " + msg, xbmc.LOGDEBUG)


Debug("Loading '%s' version '%s'" % (__scriptname__, __version__))


# check exclusion settings for filename passed as argument
def isExcluded(movie_full_path):
    if not movie_full_path:
        return False

    Debug("isExcluded(): Checking exclusion settings for '%s'." % movie_full_path)

    if (movie_full_path.find("pvr://") > -1) and __addon__.getSettingBool('ExcludeLiveTV'):
        Debug("isExcluded(): Video is playing via Live TV, which is currently set as excluded location.")
        return False

    if (movie_full_path.find("http://") > -1 or movie_full_path.find("https://") > -1) \
            and __addon__.getSettingBool('ExcludeHTTP'):
        Debug("isExcluded(): Video is playing via HTTP source, which is currently set as excluded location.")
        return False

    ExcludePath = __addon__.getSetting('ExcludePath')
    if ExcludePath and __addon__.getSettingBool('ExcludePathOption'):
        if movie_full_path.find(ExcludePath) > -1:
            Debug("isExcluded(): Video is playing from '%s', which is currently set as excluded path 1." % ExcludePath)
            return False

    ExcludePath2 = __addon__.getSetting('ExcludePath2')
    if ExcludePath2 and __addon__.getSettingBool('ExcludePathOption2'):
        if movie_full_path.find(ExcludePath2) > -1:
            Debug("isExcluded(): Video is playing from '%s', which is currently set as excluded path 2." % ExcludePath2)
            return False

    ExcludePath3 = __addon__.getSetting('ExcludePath3')
    if ExcludePath3 and __addon__.getSettingBool('ExcludePathOption3'):
        if movie_full_path.find(ExcludePath3) > -1:
            Debug("isExcluded(): Video is playing from '%s', which is currently set as excluded path 3." % ExcludePath3)
            return False

    ExcludePath4 = __addon__.getSetting('ExcludePath4')
    if ExcludePath4 and __addon__.getSettingBool('ExcludePathOption4'):
        if movie_full_path.find(ExcludePath4) > -1:
            Debug("isExcluded(): Video is playing from '%s', which is currently set as excluded path 4." % ExcludePath4)
            return False

    ExcludePath5 = __addon__.getSetting('ExcludePath5')
    if ExcludePath5 and __addon__.getSettingBool('ExcludePathOption5'):
        if movie_full_path.find(ExcludePath5) > -1:
            Debug("isExcluded(): Video is playing from '%s', which is currently set as excluded path 5." % ExcludePath5)
            return False

    return True


class AutoSubsPlayer(xbmc.Player):
    def __init__(self, *args, **kwargs):
        xbmc.Player.__init__(self)
        Debug("Initialized")
        self.run = True

    def onPlayBackStopped(self):
        Debug("Stopped")
        self.run = True

    def onPlayBackEnded(self):
        Debug("Ended")
        self.run = True

    def onPlayBackStarted(self):
        if self.isPlayingVideo():
            check_for_specific = __addon__.getSettingBool('check_for_specific')
            specific_language = __addon__.getSetting('selected_language')
            Debug("Specific language from settings '%s'" % specific_language)
            specific_language = xbmc.convertLanguage(specific_language, xbmc.ISO_639_2)
            Debug("Specific language ISO_639_2 '%s'" % specific_language)
            try:
                monitor.waitForAbort(3)
                if self.getSubtitles():
                    Debug("Subtitles already present '%s'" % self.getSubtitles())
                    self.run = False
            except:
                pass
            if self.run:
                movie_full_path = self.getPlayingFile()
                Debug("movie_full_path '%s'" % movie_full_path)
                xbmc.sleep(1000)
                available_langs = self.getAvailableSubtitleStreams()
                Debug("available_langs '%s'" % available_langs)
                total_time = self.getTotalTime()
                Debug("total_time '%s'" % total_time)
                video_clip_album = ''
                if __addon__.getSettingBool('ExcludeVideoClip'):
                    video_clip_album = xbmc.InfoTagMusic.getAlbum()
                    Debug("videoclip_album '%s'" % video_clip_album)

                if (total_time > ExcludeTime and (not video_clip_album) and (
                        (not xbmc.getCondVisibility("VideoPlayer.HasSubtitles"))
                        or (check_for_specific and specific_language not in available_langs))
                        and all(movie_full_path.find(v) <= -1 for v in ignore_words)
                        and (isExcluded(movie_full_path))):
                    self.run = False
                    xbmc.sleep(1000)
                    Debug('Started: AutoSearching for Subs')
                    xbmc.executebuiltin('ActivateWindow(SubtitleSearch)')
                else:
                    Debug('Started: Subs found or Excluded')
                    self.showSubtitles(True)
                    self.run = False
        else:
            Debug('Started: Not a video file, finishing')


class AutoSubsRunner:
    player = AutoSubsPlayer()
    while not monitor.abortRequested():
        monitor.waitForAbort(1)

    del player


default.py
#!/usr/bin/python

import os
import xbmc
import xbmcaddon

__addon__ = xbmcaddon.Addon()
__author__ = __addon__.getAddonInfo('author')
__scriptid__ = __addon__.getAddonInfo('id')
__scriptname__ = __addon__.getAddonInfo('name')
__version__ = __addon__.getAddonInfo('version')
__language__ = __addon__.getLocalizedString
debug = __addon__.getSetting("debug")
__cwd__ = xbmc.translatePath(__addon__.getAddonInfo('path'))
__profile__ = xbmc.translatePath(__addon__.getAddonInfo('profile'))
__resource__ = xbmc.translatePath(os.path.join(__cwd__, 'resources'))

__settings__ = xbmcaddon.Addon("service.autosubs")

ignore_words = (__settings__.getSetting('ignore_words').split(','))
ExcludeTime = int((__settings__.getSetting('ExcludeTime'))) * 60

currentPlaying = ""

monitor = xbmc.Monitor()


def Debug(msg, force=False):
    if (debug == "true" or force):
        xbmc.log("#####[AutoSubs]##### " + msg, xbmc.LOGDEBUG)

Debug("Loading '%s' version '%s'" % (__scriptname__, __version__))


# check exclusion settings for filename passed as argument
def isExcluded(movieFullPath):
    if not movieFullPath:
        return False

    Debug("isExcluded(): Checking exclusion settings for '%s'." % movieFullPath)

    if (movieFullPath.find("pvr://") > -1) and __addon__.getSettingBool('ExcludeLiveTV'):
        Debug("isExcluded(): Video is playing via Live TV, which is currently set as excluded location.")
        return False

    if (movieFullPath.find("http://") > -1 or movieFullPath.find("https://") > -1) and __addon__.getSettingBool('ExcludeHTTP'):
        Debug("isExcluded(): Video is playing via HTTP source, which is currently set as excluded location.")
        return False

    ExcludePath = __addon__.getSetting('ExcludePath')
    if ExcludePath and __addon__.getSettingBool('ExcludePathOption'):
        if (movieFullPath.find(ExcludePath) > -1):
            Debug("isExcluded(): Video is playing from '%s', which is currently set as excluded path 1." % ExcludePath)
            return False

    ExcludePath2 = __addon__.getSetting('ExcludePath2')
    if ExcludePath2 and __addon__.getSettingBool('ExcludePathOption2'):
        if (movieFullPath.find(ExcludePath2) > -1):
            Debug("isExcluded(): Video is playing from '%s', which is currently set as excluded path 2." % ExcludePath2)
            return False

    ExcludePath3 = __addon__.getSetting('ExcludePath3')
    if ExcludePath3 and __addon__.getSettingBool('ExcludePathOption3'):
        if (movieFullPath.find(ExcludePath3) > -1):
            Debug("isExcluded(): Video is playing from '%s', which is currently set as excluded path 3." % ExcludePath3)
            return False

    ExcludePath4 = __addon__.getSetting('ExcludePath4')
    if ExcludePath4 and __addon__.getSettingBool('ExcludePathOption4'):
        if (movieFullPath.find(ExcludePath4) > -1):
            Debug("isExcluded(): Video is playing from '%s', which is currently set as excluded path 4." % ExcludePath4)
            return False

    ExcludePath5 = __addon__.getSetting('ExcludePath5')
    if ExcludePath5 and __addon__.getSettingBool('ExcludePathOption5'):
        if (movieFullPath.find(ExcludePath5) > -1):
            Debug("isExcluded(): Video is playing from '%s', which is currently set as excluded path 5." % ExcludePath5)
            return False

    return True


class AutoSubsPlayer(xbmc.Player):
    def __init__(self, *args, **kwargs):
        xbmc.Player.__init__(self)
        Debug("Initalized")
        self.run = True

    def onPlayBackStopped(self):
        Debug("Stopped")
        self.run = True

    def onPlayBackEnded(self):
        Debug("Ended")
        self.run = True

    def onPlayBackStarted(self):
        if (self.isPlayingVideo()):
            check_for_specific = (__addon__.getSetting('check_for_specific').lower() == 'true')
            specific_language = (__addon__.getSetting('selected_language'))
            specific_language = xbmc.convertLanguage(specific_language, xbmc.ISO_639_2)
            try:
                monitor.waitForAbort(3)
                if self.getSubtitles():
                    Debug("Subtitles already present '%s'" % self.getSubtitles())
                    self.run = False
            except:
                pass
            if self.run:
                movieFullPath = self.getPlayingFile()
                Debug("movieFullPath '%s'" % movieFullPath)
                xbmc.sleep(1000)
                availableLangs = self.getAvailableSubtitleStreams()
                Debug("availableLangs '%s'" % availableLangs)
                totalTime = self.getTotalTime()
                Debug("totalTime '%s'" % totalTime)
                videoclipAlbum = ''
                if __addon__.getSettingBool('ExcludeVideoClip'):
                    videoclipAlbum = xbmc.InfoTagMusic.getAlbum()
                    Debug("videoclipAlbum '%s'" % videoclipAlbum)

                if (totalTime > ExcludeTime and (not videoclipAlbum) and (
                    (not xbmc.getCondVisibility("VideoPlayer.HasSubtitles")) or (
                    check_for_specific and not specific_language in availableLangs)) and all(
                            movieFullPath.find(v) <= -1 for v in ignore_words) and (isExcluded(movieFullPath))):
                    self.run = False
                    xbmc.sleep(1000)
                    Debug('Started: AutoSearching for Subs')
                    xbmc.executebuiltin('ActivateWindow(SubtitleSearch)')
                else:
                    Debug('Started: Subs found or Excluded')
                    self.showSubtitles(True)
                    self.run = False
        else:
            Debug('Started: Not a video file, finishing')


player = AutoSubsPlayer()

while not monitor.abortRequested():
    if player.isPlaying():
        if (not xbmc.getCondVisibility("VideoPlayer.HasSubtitles")):
            monitor.waitForAbort(3)
            if player.isPlayingVideo() and player.getPlayingFile() != currentPlaying:
                xbmc.executebuiltin('ActivateWindow(SubtitleSearch)')
                currentPlaying = player.getPlayingFile()
    monitor.waitForAbort(1)

del player
